Source/Core/Core/WC24PatchEngine.cpp

```diff
@@ -0,0 +1,158 @@
+
+  if (items.size() != 3)
+    return false;
+
+  patch->source = items[0];
+  patch->replacement = items[1];
+
+  if (!TryParse(items[2], &patch->is_kd))
+    return false;
+
+  return patch;
+}
+
+void LoadPatchSection(const Common::IniFile& ini)
+{
+  std::vector<std::string> lines;
+  NetworkPatch patch;
+  ini.GetLines("WC24Patch", &lines);
+
+  for (std::string& line : lines)
+  {
+    if (line.empty())
+      continue;
+
+    if (line[0] == '$')
+    {
+      patch.name = line.substr(1, line.size() - 1);
+    }
+    else
+    {
+      if (DeserializeLine(line, &patch))
+      {
+        s_patches.push_back(patch);
+      }
+    }
+  }
+
+  ReadEnabledAndDisabled(ini, "WC24Patch", &s_patches);
+}
+
+void LoadPatches()
+{
+  const auto& sconfig = SConfig::GetInstance();
+  // We can only load WC24 Channels.
+  if (!IsWC24Channel())
+    return;
+
+  Common::IniFile ini;
+  // If WiiLink is enabled then we load the Default Ini as that has the WiiLink URLs.
+  if (Config::Get(Config::MAIN_WII_WIILINK_ENABLE))
+    ini = sconfig.LoadDefaultGameIni();
+  else
+    ini = sconfig.LoadLocalGameIni();
+
+  LoadPatchSection(ini);
+}
+
+bool IsWC24Channel()
+{
+  const auto& sconfig = SConfig::GetInstance();
+  const auto found =
+      std::find(s_wc24_channels.begin(), s_wc24_channels.end(), sconfig.GetTitleID());
+
+  return found != s_wc24_channels.end();
+}
+
+void Reload()
+{
+  s_patches.clear();
+  LoadPatches();
+}
+
+std::optional<std::string> GetNetworkPatch(const std::string& source, IsKD is_kd)
+{
+  const auto patch =
+      std::find_if(s_patches.begin(), s_patches.end(), [&source, &is_kd](NetworkPatch& patch) {
+        return patch.source == source && patch.is_kd == is_kd;
+      });
+  if (patch == s_patches.end())
+    return std::nullopt;
+
+  return patch->replacement;
+}
+
+// When we patch for the Socket, we aren't given the URL. Instead, it is in a Host header
+// in the payload that the socket is going to send. We need to manually find which patch to apply.
+std::optional<std::string> GetNetworkPatchByPayload(std::string_view source)
+{
+  size_t pos{};
+  while (pos < source.size())
+  {
+    const size_t end_of_line = source.find("\r\n", pos);
+    if (source.substr(pos).starts_with("Host: "))
+    {
+      const std::string_view domain =
+          source.substr(pos + 6, end_of_line == std::string_view::npos ? std::string_view::npos :
+                                                                         (end_of_line - pos - 5));
+      for (const WC24PatchEngine::NetworkPatch& patch : s_patches)
+      {
+        if (patch.is_kd != WC24PatchEngine::IsKD{true} && domain == patch.source)
+        {
+          return fmt::format("{}Host: {}{}", source.substr(0, pos), patch.replacement,
+                             end_of_line == std::string_view::npos ? "" :
+                                                                     source.substr(end_of_line));
+        }
+      }
+    }
```

Comment:
Once the Host header is found but none of the patches match you can break out of the loop and return std::nullopt.

Reply:
Oh and I just realized... the patches have an `enabled` field that isn't checked here, I assume it should be?